src/units/abilities.cpp

```diff
@@ -885,33 +879,118 @@ std::string attack_type::weapon_specials(bool only_active, bool is_backstab) con
 	}
 	std::string weapon_abilities;
 	std::set<std::string> checking_name;
-	assert(display::get_singleton());
-	const unit_map& units = display::get_singleton()->get_units();
 	if(self_){
-		for (const config::any_child sp : self_->abilities().all_children_range()){
-			const bool active = check_self_abilities_impl(shared_from_this(), other_attack_, sp.cfg, self_, self_loc_, AFFECT_SELF, sp.key);
-			add_name(weapon_abilities, active, sp, checking_name, false);
+		weapon_specials_impl_self(weapon_abilities, self_, shared_from_this(), other_attack_, self_loc_, AFFECT_SELF, checking_name);
+		weapon_specials_impl_adj(weapon_abilities, self_, shared_from_this(), other_attack_, self_loc_, AFFECT_SELF, checking_name);
+	}
+	if(other_){
+		weapon_specials_impl_self(weapon_abilities, other_, other_attack_, shared_from_this(), other_loc_, AFFECT_OTHER, checking_name);
+		weapon_specials_impl_adj(weapon_abilities, other_, other_attack_, shared_from_this(), other_loc_, AFFECT_OTHER, checking_name);
+	}
+	if(!weapon_abilities.empty() && !res.empty()) {
+		weapon_abilities = ", \n" + weapon_abilities;
+		res += weapon_abilities;
+	} else if (!weapon_abilities.empty()){
+		res = weapon_abilities;
+	}
+	return res;
+}
+
+static void add_string(std::string& weapon_abilities, std::string& self_abilities, const std::string temp_abilities, std::set<std::string>& checking_name)
+{
+	weapon_abilities = "";
+	checking_name.clear();
+	self_abilities += (!self_abilities.empty() && !temp_abilities.empty()) ? "\n" : "";
+	self_abilities += temp_abilities;
+}
+
+std::string attack_type::weapon_specials_value(const std::set<std::string> checking_tags) const
+{
+	//log_scope("weapon_specials_value");
+	std::string weapon_abilities, self_abilities, self_teacher_allies, self_teacher_enemies, opponent_abilities;
+	std::set<std::string> checking_name;
+	for (const config::any_child sp : specials_.all_children_range()) {
+		if((checking_tags.count(sp.key) != 0)){
+			const bool active = special_active(sp.cfg, AFFECT_SELF, sp.key);
+			add_name(weapon_abilities, active, sp.cfg["name"].str(), checking_name);
```

Comment:
I can't follow the logic of this function - it seems that `weapon_abilities` is a temporary string that's used for assembling parts of other strings called `self_abilities` etc, and if that's what it is then I don't see why `weapon_abilities` is in scope for the whole function. I also can't see why it's an input parameter to the weapon_specials_*_self functions, as it seems to be an output from those.

However, this line uses it purely as an output, and IIUC the value set here would be discarded, were it not for the first weapon_specials_impl_self call taking it as an input, and then possibly concatenating other values to it before copying it to self_abilities. However, that happens in a conditional block that only happens if self_ is non-null (which might always be true, but then why is there a conditional block)?

Reply:
weapon_abilities used mike reference,in add_name, i cannot put other stinrg in input or i put the code itself. I will try ths option.